### ISSUE TITLE
run_maintenance drained all IO-credits and stalled my DB server

### ISSUE THREAD

**Issue by berlic** [pgpartman/pg_partman#213]
Hi,

It is not something I'm very angry about, but just want it to be here for possible future seekers.

We have just upgraded our PostgreSQL from 9.3 to 9.6 and also upgraded pg_partman from ancient 1.4.3 to 2.6.4 (latest pre-3 version).
During nightly `run_maintenance` cron-job our server started to produce heavy IO-load that drained our EBS IO-burst credits and made our server unusable.
Looking thought the logs I discovered that pg_partman has been creating numerous partitions despite `premake=3` setting.

Further digging showed that this version of `pg_partman` makes decision to premake partitions based on `max(datetime)+premake` and not `current_date+premake`. And because of the fact that we have some malformed rows in the parent table (one of which happened to have `2097-01-01` date) `pg_partman` decided to create 1000 new partitions.

This may be reasonable design decision to catch up if one forget to run `run_maintenance`, but it can lead to uncontrollable premake if occasional "future" messages possible.

But what about heavy IO-load? When I was looking through the code of `run_maintenance` to disable this `max(datetime)` behaviour, I bumped into `ANALYZE` inside(!) partition creation step. This way in my case of 1000 new partitions, analyze step was on it's way to be executed 1000 times instead of one time after all partitions were made. This is what stalled my server in the end. Am I right?

**Comment by keithf4** [comment]
What you described sounds almost like what may have happened. However, each of the things you ran into is configurable if you want to control it. 

The option "infinite_time_partitions" causes partman to base the new partition on CURRENT_DATE vs the max value in the partition set. It doesn't do this to avoid malformed future data getting entered, but to cause partman to continue to keep making future time partitions even when no new data is getting entered. Hence the name of the option. There is no option like this for serial/integer based partitioning and it's always based on the max value. Not sure of any other way to do it there.

Also, partman does do an analyze whenever it creates new children. However it only does that after the block of new child tables that are provided to the creation function have been created. So it only should've run the analyze once if all of these new child tables were created at once. So I'm not sure what happened there. An analyze is required at some point on the parent of a partition set when new children are created to keep the planner up to date so that constraint exclusion works (this may not be an issue with native partitioning in the future thankfully). You can turn off the automatic analyze in partman if you wish, by setting the "p_analyze" parameter to "run_maintenance() to false. But then you'll have to manually schedule one at some point since autovacuum will almost never kick in on the parent table since it never gets writes (it will only kick in when autovacuum_freeze_max_age is hit, which is rare and ideally should rarely happen).

And yes, if partman creates a new partition and the intervening ones are missing at that time, it will go back and create all the missing ones. This was an often requested feature during partman's early days and has since become something people find very, very convenient. However, it the newest partition was already made and previous, intervening partitions are removed, it does not then go back and make them again. It only does that when the newest partition created would cause a gap.

I do apologize for what happened, but I cannot really account for any and all forms of malformed data getting thrown into a partition set. Especially when designing features that are supposed to make things easier for the most common case. If you're concerned about it happening again, I suggest changing the two settings I mentioned.

**Comment by berlic** [comment]
`infinite_time_partitions` is taken into account only when there's no data, according to:
https://github.com/keithf4/pg_partman/blob/v2.6.4/sql/functions/run_maintenance.sql#L165-L180

Whereas in my case there are some messages with `2097-01-01` date in parent table (this bad date from distant future is in no way related to `pg_partman` operation, but because of broken time-sync on some of devices that send messages to my DB) – thus `v_current_partition_timestamp` is set by code:

```
        IF v_max_time_parent > v_current_partition_timestamp THEN
            SELECT suffix_timestamp INTO v_current_partition_timestamp FROM @extschema@.show_partition_name(v_row.parent_table, v_max_time_parent::text);
        END IF;
```

So `infinite_time_partitions` is not an option in my case, because there are occasional messages from distant future, and I'm OK to have several rows in my parent table instead of 1000 empty premade partitions.

As for `ANALYZE`: from the [code](https://github.com/keithf4/pg_partman/blob/v2.6.4/sql/functions/run_maintenance.sql#L223) I understand that `create_partition_time` is called for every new partition to be created, and `ANALYZE` is called from inside it [here](https://github.com/keithf4/pg_partman/blob/v2.6.4/sql/functions/create_partition_time.sql#L300). So I'm not quite sure about your words:
> However it only does that after the block of new child tables that are provided to the creation function have been created

From the logs:

```
...
Creating partition function for table xxx.yyy
Creating new partition xxx.yyy_p2021_03
Setting new child table privileges for xxx.yyy_p2021_03
Applying foreign keys to xxx.yyy_p2021_03
Analyzing partition set: xxx.yyy
Creating partition function for table xxx.yyy
Creating new partition xxx.yyy_p2021_04
Setting new child table privileges for xxx.yyy_p2021_04
Applying foreign keys to xxx.yyy_p2021_04
Analyzing partition set: xxx.yyy
...
```

So it analyzed after every partition and not after all 1000 has been made.

In any case I do understand that version 2.6.4 is already 1-year-old and things may have changed in version 3.x.

Despite this minor incident, thank you for the product! It helped us to maintain massive amount of incoming data in PostgreSQL!

**Comment by keithf4** [comment]
So I hadn't closed this issue because I wanted to see if I could find where the analyze is getting thrown in there unnecessarily, but I haven't been able to narrow it down. However, I'm going to be doing a bit of an overhaul on maintenance in general with the upcoming PROCEDURE feature in PG11

https://github.com/pgpartman/pg_partman/issues/219

I believe this is going to solve many of the lingering maintenance issues in pg_partman since right now it's stuck doing everything in one giant transaction. The work for this will be getting done over the summer, so it will be ready when PG11 is released. Hopefully this maintenance review will catch where this extra load is being caused.

**Comment by keithf4** [comment]
After reviewing the code I'm still at a bit of a loss for how this happened. The analyzes shouldn't be being done until after all of the child tables are created for a set in a single maintenance run. Unless what you pulled from the log was from distinct maintenance runs about a month apart? For monthly partitioning, only one partition is going to be being created per month under normal maintenance situations, so seeing an analyze run after each child table is normal. 

I've released version 4.0.0 which has a procedure version of run_maintenance available in PG11 (which should hopefully be out this month). This should greatly lessen contention during maintenance since it commits after each partition set's maintenance is complete. It also no longer analyzes by default for native sets on PG11.

If you're still seeing these sorts of IO issues with version 4, please let me know. I'll be closing this issue for now, but please feel free to reply and request it be reopened at any time.
